src/Profilers/MonitorProfiler/Communication/MessageCallbackManager.cpp

```diff
@@ -27,7 +36,7 @@ bool MessageCallbackManager::TryRegister(unsigned short commandSet, std::functio
 
 HRESULT MessageCallbackManager::DispatchMessage(const IpcMessage& message)
 {
-    std::lock_guard<std::mutex> lock(m_mutex);
+    std::lock_guard<std::mutex> dispatchLock(m_dispatchMutex);
 
```

Comment:
Shouldn't this also acquire a lock on `m_lookupMutex` since it is accessing the callback list? I wonder if the guard should be moved into `TryGetCallback` so then the callers don't have to worry about getting the lock.

Reply:
So this is where a `shared_mutex` would make things more obvious as to the intent. 

I'm (poorly) replicating support for multiple readers by having 1 mutex for each distinct reader:
- 1 thread will be repeatedly calling `DispatchMessage`
- 1 thread will be repeatedly calling `IsRegistered`

These 2 operations need to be able to occur at the same time, since `DispatchMessage` can take a long time and we don't want to hold up the other thread that calls `IsRegistered`.

Meanwhile N threads can be calling into either `TryRegister` or `Unregister`. These are the "writers" and so they acquire both mutexes to guarantee exclusivity. 

----

> I wonder if the guard should be moved into TryGetCallback so then the callers don't have to worry about getting the lock.

`TryGetCallback` is a helper method used by both "readers" and "writers" which require holding the locks for different scopes so that's why we don't grab a lock inside of there.

Reply:
So this is where a shared_mutex would make things more obvious as to the intent.

I'm (kinda) replicating support for multiple readers by having 1 mutex for each distinct reader:

1 thread will be repeatedly calling DispatchMessage
1 thread will be repeatedly calling IsRegistered
These 2 operations need to be able to occur at the same time, since DispatchMessage can take a long time and we don't want to hold up the other thread that calls IsRegistered.

Meanwhile N threads can be calling into either TryRegister or Unregister. These are the "writers" and so they acquire both mutexes to guarantee exclusivity.

I wonder if the guard should be moved into TryGetCallback so then the callers don't have to worry about getting the lock.

TryGetCallback is a helper method used by both "readers" and "writers" which require holding the locks for different scopes so that's why we don't grab a lock inside of there.